packages/grafana-ui/src/components/Segment/SegmentInput.tsx

```diff
@@ -0,0 +1,51 @@
+import React, { useRef, useState } from 'react';
+import { cx, css } from 'emotion';
+import useClickAway from 'react-use/lib/useClickAway';
+import { useExpandableLabel, SegmentProps } from '.';
+
+export interface SegmentInputProps<T> extends SegmentProps<T> {
+  value: string | number;
+  onChange: (text: string | number) => void;
+}
+
+const textWidth = (text: string) => {
+  const element = document.createElement('canvas');
+  const context = element.getContext('2d')!;
+  context.font = '14px Roboto';
+  return context.measureText(text).width;
+};
+
+export function SegmentInput<T>({
+  value,
+  onChange,
+  Component,
+  className,
+}: React.PropsWithChildren<SegmentInputProps<T>>) {
+  const ref = useRef(null);
+  const [inputWidth, setInputWidth] = useState<number>(textWidth(value.toString()));
+  const [Label, , expanded, setExpanded] = useExpandableLabel(false);
+  useClickAway(ref, () => setExpanded(false));
+
+  if (!expanded) {
+    return <Label Component={Component || <a className={cx('gf-form-label', 'query-part', className)}>{value}</a>} />;
+  }
+
+  const inputWidthStyle = css`
+    width: ${Math.max(inputWidth + 20, 32)}px;
+  `;
+
+  return (
+    <input
+      ref={ref}
+      autoFocus
+      className={cx(`gf-form gf-form-input`, inputWidthStyle)}
+      value={value}
+      onChange={item => {
+        setInputWidth(textWidth(item.target.value));
```

Comment:
I'm not super wild about this as this creates the canvas element on every change. 

Two alternative solutions that come to my mind:
1. Rendering an invisible div next to the input. This div would have exactly the same style as the input element applied, and also would have the same text as the input. Then you could use ref to this element, measure it and apply measured width to the input .

2. Render this div as a transparent one and make it "push" the size of the container. The input would have to be positioned absolutely within the container and have `width: 100%;` set.


In general mi thinking about the markup for this is:

```
<div container>
  <input ... />
  <div sizeReference />
</div>
```

Reply:
We have a util function to measure text width, https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/utils/measureText.ts#L4 

it has both a measure cache & canvas cache that are reused.  Think with that this approach should be fine?

Reply:
Good suggestions @dprokop! 

> it has both a measure cache & canvas cache that are reused. Think with that this approach should be fine?
Okay, did not know that! I'll do that instead. I think it will be sufficient.